WordPress/openverse-api
# Add throttle exemptions

## Fixes


Fixes #565 by @zackkrida 

## Description


Adds the ability to easily define exemptions for throttle classes via a general DRF/Django OOP inspired approach.

The idea is that each exemption is defined as its own class which implements the `is_exempt` method which returns a boolean: true when exempt and false otherwise. If any exemption class declares the request as exempt from throttling, then the throttle's `allow_request` logic is short circuited and the request is allowed.

I've added comprehensive unit tests for this. I tried to cover every possible configuration and request situation with redundant checks against the pre-existing state before checking the requests.

The tests are roughly modeled after DRF's own tests for the base throttling classes: https://github.com/encode/django-rest-framework/blob/master/tests/test_throttling.py.

## Testing Instructions

As I mentioned, the unit tests should cover the code 100%.

We can also rely on the integration tests, they should all continue to pass without exception.

Smoke testing the API by running it locally and checking a few requests wouldn't hurt either. If you really feel like it, modifying the local running redis instance to add localhost IP or an API key to the client id allow list would also be good.

For example, on my computer, I exec'd into the redis container using:

```bash
docker-compose exec cache redis-cli
```

Then ran:

```
SADD 'ip-whitelist' '172.20.0.1'
```

And verified that I would not get rate limited.

Now run:

```
SREM 'ip-whitelist' '172.20.0.1'
```

and confirm that you _do_ get throttled.

I think you could do the same using the `AccessTokenFactory` in `just ipython` but there's some extra configuration you need to do to get the locally running app configured for oauth I guess.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
